### PR TITLE
Upgrade Gradle 8.4 to 8.5 and upgrade related Gradle Plugins

### PR DESCRIPTION
**Pull request by TWiStErRob**
See individual commits. Let me know how this should be done if this YOLO PR is too broad.

Proposed commit message:

```
Upgrade Gradle 8.4 to 8.5 and upgrade related Gradle Plugins

Gradle 8.5 brings performance updates and full Java 21 support.
Some plugins are already incompatible with Gradle 8.4,
so updating them to be in line with the latest Gradle release.
Minor changes:
 * Add .gitattributes lines for Gradle files,
to ensure they're not clobbered during git operations.
 * Remove `com.google.jimfs:jimfs`'s automatic module name,
because it's already defined and errors after bumping extra-java-module-info.
 * Added property to tell Sonar to work in a future-proof way.
 * Removed unused :demo project.

PR: #805
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.adoc#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [ ] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [ ] Site documentation in `.adoc` file references demo in `src/demo/java` instead of containing code blocks as text
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [ ] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [ ] The `package-info.java` contains information about the new extension

Code (general)
* [ ] Code adheres to code style, naming conventions etc.
* [ ] Successful tests cover all changes
* [ ] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [ ] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.adoc#assertions) (which are based on AssertJ)

Code (new package)
* [ ] The new package is exported in `module-info.java`
* [ ] The new package is also present in the tests
* [ ] The new package is opened for reflection to JUnit 5 in `module-info.java`
* [ ] The new package is listed in the contribution guide

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.adoc` mentions the new contribution (real name optional)

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -198,11 +197,6 @@ tasks {
 			runtimeClasspath += sourceSets.main.get().output
 		}
 	}
-	project(":demo") {
-		this.sonar {
-			isSkipProject = true
-		}
-	}
```
Comment (Michael1993) [inline build.gradle.kts:205]: Why did you delete this part?

File: settings.gradle.kts [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 	}
-}
\ No newline at end of file
+}
+
+// See https://community.sonarsource.com/t/sonar-gradle-skipcompile-is-not-working/102710/9
+// TODO move this to a better place once https://sonarsource.atlassian.net/browse/SONARGRADL-134 is resolved.
+// > The 'sonar' / 'sonarqube' task depends on compile tasks.
+// > This behavior is now deprecated and will be removed in version 5.x.
+// > To avoid implicit compilation, set property 'sonar.gradle.skipCompile' to 'true' and make sure your project is compiled, before analysis has started.
```
Comment (Michael1993) [inline settings.gradle.kts:18]: This needs to be a proper issue if/when we merge this, I think.

File: build.gradle.kts [hunk 3]
```diff
@@ -4,14 +4,14 @@ plugins {
...
+	id("com.diffplug.spotless") version "6.25.0"
 	id("at.zierler.yamlvalidator") version "1.5.0"
-	id("org.sonarqube") version "4.3.1.3277"
-	id("org.shipkit.shipkit-changelog") version "1.2.0"
-	id("org.shipkit.shipkit-github-release") version "1.2.0"
-	id("com.github.ben-manes.versions") version "0.48.0"
-	id("io.github.gradle-nexus.publish-plugin") version "1.3.0"
-	id("org.gradlex.extra-java-module-info") version "1.4.2"
+	id("org.sonarqube") version "4.4.1.3373"
+	id("org.shipkit.shipkit-changelog") version "2.0.1"
```
Comment (Michael1993) [inline build.gradle.kts:10]: This is a major version change. Are you sure the project is compatible?

File: .gitattributes [hunk 4]
```diff
@@ -5,3 +5,6 @@
 # force all text files on the working dir to have LF line endings
 # https://git-scm.com/docs/gitattributes#_eol
 * text eol=lf
+
+*.bat text eol=crlf
+*.jar binary
```
Comment (Michael1993) [inline .gitattributes:10]: Not sure what this does...?

**Review by Michael1993 — COMMENTED** [review]
Thank you for this! Could you open an issue where we can discuss all your different changes and PRs with the other maintainers?

File: .gitattributes (hunk 4 above)
Comment (TWiStErRob) [reply .gitattributes:10]: It tells git to not touch the line endings.

`.bat` files are meant for Windows, so they should end in CRLF, currently they're LF, because of the top wildcard.

`.jar` files are `binary` file, meaning they are not `text`, line endings (`eol`) are meaningless for binary files. When I committed the new wrapper, it said the next time I do a git operation it'll replace line endings, which means the `.jar` contents would change and the `ZIP` decoder could/would fail.

These two are quite standard and well known in Gradle repos:
https://www.google.com/search?q=*.bat+text+eol%3Dcrlf+*.jar+binary

File: build.gradle.kts (hunk 3 above)
Comment (TWiStErRob) [reply build.gradle.kts:10]: Looking through the changes (not much) the ONLY relevant change for consumers is that they're building with Gradle 7 now: https://github.com/shipkit/shipkit-changelog/compare/v1.2.0...v2.0.1#diff-40640fe1078ece83d7ea8fb67daacd77923a86d13447baf9769660b3b46f2eceR3. I think they bumped major because maybe they think building with Gradle 7 makes minimum Gradle 7. Or it's a mono-release-train within the Shipkit org and something else was bumped higher.

Note: The Gradle task input fix was done in 1.2.3.

File: build.gradle.kts (hunk 1 above)
Comment (TWiStErRob) [reply build.gradle.kts:205]: See https://github.com/junit-pioneer/junit-pioneer/pull/805/commits/299c95ef51c0737c9bebd757981973d02e7586a4

Comment (TWiStErRob) [reply build.gradle.kts:205]: Is the `:demo` project used in any way? I couldn't find any scripts or references to it in the repo other than definition and exclusion from sonar.

File: settings.gradle.kts (hunk 2 above)
Comment (TWiStErRob) [reply settings.gradle.kts:18]: Do you mean create an issue for the TODO to be fixed? (it'll linger for a long time, as there's no progress on it)

The warning (`>` text) is solved, the CI is doing a full build before running sonar:

https://github.com/junit-pioneer/junit-pioneer/blob/b02cda658464d548c011a1efb651cbf875a2dea4/.github/workflows/build.yml#L95-L103

**Comment by TWiStErRob** [comment]
> Could you open an issue where we can discuss all your different changes and PRs with the other maintainers?

I'm not sure how to action this. The umbrella would be "CI uses outdated tools", but that might also increase the scope. These are 3 small isolated updates (similar to what Dependabot/Renovate would do + fixing deprecations) each target what's in the title.

**Comment by Bukama** [comment]
Hey, thanks for the PR. @Michael1993 already reviewed it. Therefore only one thing by me:

>  * Removed unused :demo project.

The demo project is used as code base for the documentation

File: settings.gradle.kts [hunk 5]
```diff
@@ -9,4 +9,11 @@ gradleEnterprise {
 		termsOfServiceUrl = "https://gradle.com/terms-of-service"
 		termsOfServiceAgree = "yes"
 	}
-}
\ No newline at end of file
+}
+
+// See https://community.sonarsource.com/t/sonar-gradle-skipcompile-is-not-working/102710/9
```
Comment (beatngu13) [inline settings.gradle.kts:14]: You can remove this line as SONARGRADL-134 already links to it.

File: build.gradle.kts [hunk 6]
```diff
@@ -4,14 +4,14 @@ plugins {
...
+	id("org.shipkit.shipkit-github-release") version "2.0.1"
+	id("com.github.ben-manes.versions") version "0.51.0"
+	id("io.github.gradle-nexus.publish-plugin") version "2.0.0-rc-1"
```
Comment (beatngu13) [inline build.gradle.kts:13]: I'd rather stay on the [latest stable](https://github.com/gradle-nexus/publish-plugin/releases) (i.e. 1.3.0) instead of using an RC.

**Review by beatngu13 — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 6 above)
Comment (TWiStErRob) [reply build.gradle.kts:13]: The breaking changes are not relevant on the rc1 for most consumers.
(I'm a maintainer on that plugin, we only released RC1 so that the contributor for Ivy could use it.).

The Ivy publishing PR brought in some breaking changes:
https://github.com/gradle-nexus/publish-plugin/releases/tag/v2.0.0-rc-1

There was a build warning fixed in one of my PRs merged (sadly) after RC1 was cut.

File: settings.gradle.kts (hunk 5 above)
Comment (TWiStErRob) [reply settings.gradle.kts:14]: will do

Comment (TWiStErRob) [reply settings.gradle.kts]: It was originally meant for keeping even after the TODO is removed, but I realize that if we can move this to a better place, the link will become invalid too, so removed.

File: build.gradle.kts (hunk 3 above)
Comment (TWiStErRob) [reply build.gradle.kts:10]: Mockito also updated it without a hitch:
https://github.com/mockito/mockito/pull/3248
and they've released since then.

**Comment by TWiStErRob** [comment]
@Michael1993 @Bukama @beatngu13 can I please get a re-review?

File: build.gradle.kts (hunk 6 above)
Comment (beatngu13) [reply build.gradle.kts:13]: Thanks for the insights! 🙏

Meanwhile, there's RC2 available:

https://github.com/gradle-nexus/publish-plugin/releases/tag/v2.0.0-rc-2

**Comment by beatngu13** [comment]
BTW feel free to add yourself and your PRs to the contributors list:

https://github.com/junit-pioneer/junit-pioneer?tab=readme-ov-file#contributors

And if I haven't said it already: thank you! 😉

**Review by Michael1993 — APPROVED** [review]
It's fine for me, please coordinate with @beatngu13 about his requests.

File: settings.gradle.kts (hunk 2 above)
Comment (TWiStErRob) [reply settings.gradle.kts]: @Michael1993 can you please clarify or resolve?

File: build.gradle.kts (hunk 1 above)
Comment (TWiStErRob) [reply build.gradle.kts:205]: @Bukama please answer https://github.com/junit-pioneer/junit-pioneer/pull/805#issuecomment-1915280798, and/or resolve this. From the code alone I don't see `:demo` usages.

File: build.gradle.kts (hunk 6 above)
Comment (TWiStErRob) [reply build.gradle.kts]: Updated, also reviewed the publishing setup and tested as much as I could without actually publishing.

**Comment by TWiStErRob** [comment]
@beatngu13 please re-review.

File: settings.gradle.kts (hunk 2 above)
Comment (Michael1993) [reply settings.gradle.kts]: Yes, I meant the `TODO` comment should be a proper issue. My feeling on `TODO` in code is that it is more likely to be forgotten than a proper issue (even though those also tend to get forgotten after a bit...)

Comment (TWiStErRob) [reply settings.gradle.kts]: Since this Todo is based on external events which may take months or never get done. I think it's enough for the next person who touches that area, as a reminder to open the issue and see what's up.
